### ISSUE TITLE
Additional dependencies required for Windows: Windows build failure (Microsoft Visual C++ 14.0 or greater is required)

### ISSUE THREAD

**Issue by ross-spencer** [iscc/iscc-core#64]
I'll have to work on this my end, although I'm not able to be admin of this particular machine. Logging for anyone who may also happen upon this issue:

* pip 22.1.2 
* Python 3.9.6   

Running: `python -m pip install iscc-core`

```
Collecting iscc-core
  Using cached iscc-core-0.2.8.tar.gz (56 kB)
  Installing build dependencies ... done
  Getting requirements to build wheel ... done
  Preparing metadata (pyproject.toml) ... done
Collecting data-url<2.0,>=1.0
  Using cached data_url-1.0.3-py3-none-any.whl
Collecting bitarray-hardbyte
  Using cached bitarray_hardbyte-2.3.8-cp39-cp39-win_amd64.whl (103 kB)
Collecting blake3
  Using cached blake3-0.3.1-cp39-none-win_amd64.whl (193 kB)
Collecting jcs<0.3.0,>=0.2.0
  Using cached jcs-0.2.1-py3-none-any.whl (7.6 kB)
Collecting pybase64
  Using cached pybase64-1.2.2-cp39-cp39-win_amd64.whl (38 kB)
Collecting pydantic[dotenv]
  Using cached pydantic-1.9.1-cp39-cp39-win_amd64.whl (2.0 MB)
Collecting more-itertools
  Using cached more_itertools-8.13.0-py3-none-any.whl (51 kB)
Collecting xxhash
  Using cached xxhash-3.0.0-cp39-cp39-win_amd64.whl (29 kB)
Collecting uvarint<2.0,>=1.2
  Using cached uvarint-1.2.0-py3-none-any.whl (4.4 kB)
Collecting loguru<0.7,>=0.6
  Using cached loguru-0.6.0-py3-none-any.whl (58 kB)
Collecting bases<0.3,>=0.2
  Using cached bases-0.2.1-py3-none-any.whl (35 kB)
Collecting base58<3.0,>=2.1
  Using cached base58-2.1.1-py3-none-any.whl (5.6 kB)
Collecting typing-validation
  Using cached typing_validation-0.0.1.post7-py3-none-any.whl (9.5 kB)
Requirement already satisfied: typing-extensions in c:\users\spencer\apps\python\lib\site-packages (from bases<0.3,>=0.2->iscc-core) (4.1.1)
Requirement already satisfied: colorama>=0.3.4 in c:\users\spencer\apps\python\lib\site-packages (from loguru<0.7,>=0.6->iscc-core) (0.4.4)
Collecting win32-setctime>=1.0.0
  Using cached win32_setctime-1.1.0-py3-none-any.whl (3.6 kB)
Collecting python-dotenv>=0.10.4
  Using cached python_dotenv-0.20.0-py3-none-any.whl (17 kB)
Building wheels for collected packages: iscc-core
  Building wheel for iscc-core (pyproject.toml) ... error
  error: subprocess-exited-with-error

  × Building wheel for iscc-core (pyproject.toml) did not run successfully.
  │ exit code: 1
  ╰─> [75 lines of output]
      A setup.py file already exists. Using it.
      C:\Users\Spencer\AppData\Local\Temp\pip-build-env-wrk8753n\overlay\Lib\site-packages\Cython\Compiler\Main.py:369: FutureWarning: Cython directive 'language_level' not set, using 2 for now (Py2). This will change in a later release! File: C:\Users\Spencer\AppData\Local\Temp\pip-install-g8z84mcf\iscc-core_94f9d89354664a52b747d9b61b700b0a\iscc_core\cdc.pxd
        tree = Parsing.p_module(s, pxd, full_module_name)
      C:\Users\Spencer\AppData\Local\Temp\pip-build-env-wrk8753n\overlay\Lib\site-packages\Cython\Compiler\Main.py:369: FutureWarning: Cython directive 'language_level' not set, using 2 for now (Py2). This will change in a later release! File: C:\Users\Spencer\AppData\Local\Temp\pip-install-g8z84mcf\iscc-core_94f9d89354664a52b747d9b61b700b0a\iscc_core\minhash.pxd
        tree = Parsing.p_module(s, pxd, full_module_name)
      C:\Users\Spencer\AppData\Local\Temp\pip-build-env-wrk8753n\overlay\Lib\site-packages\Cython\Compiler\Main.py:369: FutureWarning: Cython directive 'language_level' not set, using 2 for now (Py2). This will change in a later release! File: C:\Users\Spencer\AppData\Local\Temp\pip-install-g8z84mcf\iscc-core_94f9d89354664a52b747d9b61b700b0a\iscc_core\simhash.pxd
        tree = Parsing.p_module(s, pxd, full_module_name)
      Compiling iscc_core/cdc.py because it changed.
      Compiling iscc_core/minhash.py because it changed.
      Compiling iscc_core/simhash.py because it changed.
      [1/3] Cythonizing iscc_core/cdc.py
      [2/3] Cythonizing iscc_core/minhash.py
      [3/3] Cythonizing iscc_core/simhash.py
      ************************************************************
      Succeeded to compile Cython/C accelerator modules :)       *
      ************************************************************
      running build
      running build_py
      creating C:\Users\Spencer\AppData\Local\Temp\pip-install-g8z84mcf\iscc-core_94f9d89354664a52b747d9b61b700b0a\build
      creating C:\Users\Spencer\AppData\Local\Temp\pip-install-g8z84mcf\iscc-core_94f9d89354664a52b747d9b61b700b0a\build\lib.win-amd64-cpython-39
      creating C:\Users\Spencer\AppData\Local\Temp\pip-install-g8z84mcf\iscc-core_94f9d89354664a52b747d9b61b700b0a\build\lib.win-amd64-cpython-39\iscc_core
      copying iscc_core\cdc.py -> C:\Users\Spencer\AppData\Local\Temp\pip-install-g8z84mcf\iscc-core_94f9d89354664a52b747d9b61b700b0a\build\lib.win-amd64-cpython-39\iscc_core
      copying iscc_core\check.py -> C:\Users\Spencer\AppData\Local\Temp\pip-install-g8z84mcf\iscc-core_94f9d89354664a52b747d9b61b700b0a\build\lib.win-amd64-cpython-39\iscc_core
      copying iscc_core\codec.py -> C:\Users\Spencer\AppData\Local\Temp\pip-install-g8z84mcf\iscc-core_94f9d89354664a52b747d9b61b700b0a\build\lib.win-amd64-cpython-39\iscc_core
      copying iscc_core\code_content_audio.py -> C:\Users\Spencer\AppData\Local\Temp\pip-install-g8z84mcf\iscc-core_94f9d89354664a52b747d9b61b700b0a\build\lib.win-amd64-cpython-39\iscc_core
      copying iscc_core\code_content_image.py -> C:\Users\Spencer\AppData\Local\Temp\pip-install-g8z84mcf\iscc-core_94f9d89354664a52b747d9b61b700b0a\build\lib.win-amd64-cpython-39\iscc_core
      copying iscc_core\code_content_mixed.py -> C:\Users\Spencer\AppData\Local\Temp\pip-install-g8z84mcf\iscc-core_94f9d89354664a52b747d9b61b700b0a\build\lib.win-amd64-cpython-39\iscc_core
      copying iscc_core\code_content_text.py -> C:\Users\Spencer\AppData\Local\Temp\pip-install-g8z84mcf\iscc-core_94f9d89354664a52b747d9b61b700b0a\build\lib.win-amd64-cpython-39\iscc_core
      copying iscc_core\code_content_video.py -> C:\Users\Spencer\AppData\Local\Temp\pip-install-g8z84mcf\iscc-core_94f9d89354664a52b747d9b61b700b0a\build\lib.win-amd64-cpython-39\iscc_core
      copying iscc_core\code_data.py -> C:\Users\Spencer\AppData\Local\Temp\pip-install-g8z84mcf\iscc-core_94f9d89354664a52b747d9b61b700b0a\build\lib.win-amd64-cpython-39\iscc_core
      copying iscc_core\code_flake.py -> C:\Users\Spencer\AppData\Local\Temp\pip-install-g8z84mcf\iscc-core_94f9d89354664a52b747d9b61b700b0a\build\lib.win-amd64-cpython-39\iscc_core
      copying iscc_core\code_instance.py -> C:\Users\Spencer\AppData\Local\Temp\pip-install-g8z84mcf\iscc-core_94f9d89354664a52b747d9b61b700b0a\build\lib.win-amd64-cpython-39\iscc_core
      copying iscc_core\code_meta.py -> C:\Users\Spencer\AppData\Local\Temp\pip-install-g8z84mcf\iscc-core_94f9d89354664a52b747d9b61b700b0a\build\lib.win-amd64-cpython-39\iscc_core
      copying iscc_core\conformance.py -> C:\Users\Spencer\AppData\Local\Temp\pip-install-g8z84mcf\iscc-core_94f9d89354664a52b747d9b61b700b0a\build\lib.win-amd64-cpython-39\iscc_core
      copying iscc_core\constants.py -> C:\Users\Spencer\AppData\Local\Temp\pip-install-g8z84mcf\iscc-core_94f9d89354664a52b747d9b61b700b0a\build\lib.win-amd64-cpython-39\iscc_core
      copying iscc_core\dct.py -> C:\Users\Spencer\AppData\Local\Temp\pip-install-g8z84mcf\iscc-core_94f9d89354664a52b747d9b61b700b0a\build\lib.win-amd64-cpython-39\iscc_core
      copying iscc_core\iscc_code.py -> C:\Users\Spencer\AppData\Local\Temp\pip-install-g8z84mcf\iscc-core_94f9d89354664a52b747d9b61b700b0a\build\lib.win-amd64-cpython-39\iscc_core
      copying iscc_core\iscc_id.py -> C:\Users\Spencer\AppData\Local\Temp\pip-install-g8z84mcf\iscc-core_94f9d89354664a52b747d9b61b700b0a\build\lib.win-amd64-cpython-39\iscc_core
      copying iscc_core\minhash.py -> C:\Users\Spencer\AppData\Local\Temp\pip-install-g8z84mcf\iscc-core_94f9d89354664a52b747d9b61b700b0a\build\lib.win-amd64-cpython-39\iscc_core
      copying iscc_core\models.py -> C:\Users\Spencer\AppData\Local\Temp\pip-install-g8z84mcf\iscc-core_94f9d89354664a52b747d9b61b700b0a\build\lib.win-amd64-cpython-39\iscc_core
      copying iscc_core\options.py -> C:\Users\Spencer\AppData\Local\Temp\pip-install-g8z84mcf\iscc-core_94f9d89354664a52b747d9b61b700b0a\build\lib.win-amd64-cpython-39\iscc_core
      copying iscc_core\simhash.py -> C:\Users\Spencer\AppData\Local\Temp\pip-install-g8z84mcf\iscc-core_94f9d89354664a52b747d9b61b700b0a\build\lib.win-amd64-cpython-39\iscc_core
      copying iscc_core\utils.py -> C:\Users\Spencer\AppData\Local\Temp\pip-install-g8z84mcf\iscc-core_94f9d89354664a52b747d9b61b700b0a\build\lib.win-amd64-cpython-39\iscc_core
      copying iscc_core\wtahash.py -> C:\Users\Spencer\AppData\Local\Temp\pip-install-g8z84mcf\iscc-core_94f9d89354664a52b747d9b61b700b0a\build\lib.win-amd64-cpython-39\iscc_core
      copying iscc_core\__init__.py -> C:\Users\Spencer\AppData\Local\Temp\pip-install-g8z84mcf\iscc-core_94f9d89354664a52b747d9b61b700b0a\build\lib.win-amd64-cpython-39\iscc_core
      copying iscc_core\cdc.c -> C:\Users\Spencer\AppData\Local\Temp\pip-install-g8z84mcf\iscc-core_94f9d89354664a52b747d9b61b700b0a\build\lib.win-amd64-cpython-39\iscc_core
      copying iscc_core\cdc.pxd -> C:\Users\Spencer\AppData\Local\Temp\pip-install-g8z84mcf\iscc-core_94f9d89354664a52b747d9b61b700b0a\build\lib.win-amd64-cpython-39\iscc_core
      copying iscc_core\data.json -> C:\Users\Spencer\AppData\Local\Temp\pip-install-g8z84mcf\iscc-core_94f9d89354664a52b747d9b61b700b0a\build\lib.win-amd64-cpython-39\iscc_core
      copying iscc_core\minhash.c -> C:\Users\Spencer\AppData\Local\Temp\pip-install-g8z84mcf\iscc-core_94f9d89354664a52b747d9b61b700b0a\build\lib.win-amd64-cpython-39\iscc_core
      copying iscc_core\minhash.pxd -> C:\Users\Spencer\AppData\Local\Temp\pip-install-g8z84mcf\iscc-core_94f9d89354664a52b747d9b61b700b0a\build\lib.win-amd64-cpython-39\iscc_core
      copying iscc_core\simhash.c -> C:\Users\Spencer\AppData\Local\Temp\pip-install-g8z84mcf\iscc-core_94f9d89354664a52b747d9b61b700b0a\build\lib.win-amd64-cpython-39\iscc_core
      copying iscc_core\simhash.pxd -> C:\Users\Spencer\AppData\Local\Temp\pip-install-g8z84mcf\iscc-core_94f9d89354664a52b747d9b61b700b0a\build\lib.win-amd64-cpython-39\iscc_core
      running build_ext
      building 'iscc_core.cdc' extension
      error: Microsoft Visual C++ 14.0 or greater is required. Get it with "Microsoft C++ Build Tools": https://visualstudio.microsoft.com/visual-cpp-build-tools/
      Traceback (most recent call last):
        File "C:\Users\Spencer\Apps\python\lib\site-packages\pip\_vendor\pep517\in_process\_in_process.py", line 363, in <module>
          main()
        File "C:\Users\Spencer\Apps\python\lib\site-packages\pip\_vendor\pep517\in_process\_in_process.py", line 345, in main
          json_out['return_val'] = hook(**hook_input['kwargs'])
        File "C:\Users\Spencer\Apps\python\lib\site-packages\pip\_vendor\pep517\in_process\_in_process.py", line 261, in build_wheel
          return _build_backend().build_wheel(wheel_directory, config_settings,
        File "C:\Users\Spencer\AppData\Local\Temp\pip-build-env-wrk8753n\overlay\Lib\site-packages\poetry\core\masonry\api.py", line 68, in build_wheel
          return unicode(WheelBuilder.make_in(poetry, Path(wheel_directory)))
        File "C:\Users\Spencer\AppData\Local\Temp\pip-build-env-wrk8753n\overlay\Lib\site-packages\poetry\core\masonry\builders\wheel.py", line 78, in make_in
          wb.build()
        File "C:\Users\Spencer\AppData\Local\Temp\pip-build-env-wrk8753n\overlay\Lib\site-packages\poetry\core\masonry\builders\wheel.py", line 110, in build
          self._build(zip_file)
        File "C:\Users\Spencer\AppData\Local\Temp\pip-build-env-wrk8753n\overlay\Lib\site-packages\poetry\core\masonry\builders\wheel.py", line 162, in _build
          self._run_build_command(setup)
        File "C:\Users\Spencer\AppData\Local\Temp\pip-build-env-wrk8753n\overlay\Lib\site-packages\poetry\core\masonry\builders\wheel.py", line 190, in _run_build_command
          subprocess.check_call(
        File "C:\Users\Spencer\Apps\python\lib\subprocess.py", line 373, in check_call
          raise CalledProcessError(retcode, cmd)
      subprocess.CalledProcessError: Command '['C:/Users/Spencer/Apps/python/python.exe', 'C:\\Users\\Spencer\\AppData\\Local\\Temp\\pip-install-g8z84mcf\\iscc-core_94f9d89354664a52b747d9b61b700b0a\\setup.py', 'build', '-b', 'C:\\Users\\Spencer\\AppData\\Local\\Temp\\pip-install-g8z84mcf\\iscc-core_94f9d89354664a52b747d9b61b700b0a\\build']' returned non-zero exit status 1.
      [end of output]

  note: This error originates from a subprocess, and is likely not a problem with pip.
  ERROR: Failed building wheel for iscc-core
Failed to build iscc-core
ERROR: Could not build wheels for iscc-core, which is required to install pyproject.toml-based projects
```

**Comment by titusz** [comment]
The installation tries to build binary extensions if Cython is available in your environment. This seems to fail due to the missing Microsoft Visual C++ 14.0 dependency. Compilation of the extension modules is optional and it should work with pure python. You could try to uninstall Cython from your environmet with `python -m pip uninstall Cython` and than try installing again with `python -m pip install iscc-core`. Let me know if that works?

**Comment by ross-spencer** [comment]
Thanks @titusz. I've tried a couple of things this morning. Cython wasn't installed locally via pip:

```
C:\Users\Spencer>python -m pip uninstall Cython
WARNING: Skipping Cython as it is not installed.
```

Just in case I created a new venv and still the same error. 

I started to research how pip was interacting with the poetry `optional` dependencies, and I can't find anything concrete on that yet. It did prompt me to download this repo and install it via poetry however. `poetry install` from root on Windows creates a venv that can be accessed. This is an okay workaround for testing the library at least.

**Comment by titusz** [comment]
I released a new version 0.2.11 on pypi that should fail gracefully if compilation does not succeed. Maybe you can give it a try.

**Comment by ross-spencer** [comment]
Nice work @titusz that works perfectly! Looking forward to trying to incorporate this some more on this platform. Thanks for this. 

```
python -m pip install iscc-core
Collecting iscc-core
  Downloading iscc-core-0.2.11.tar.gz (57 kB)
     ---------------------------------------- 57.7/57.7 kB 601.2 kB/s eta 0:00:00
  Installing build dependencies ... done
  Getting requirements to build wheel ... done
  Preparing metadata (pyproject.toml) ... done
Collecting base58<3.0,>=2.1
  Using cached base58-2.1.1-py3-none-any.whl (5.6 kB)
Collecting more-itertools
  Using cached more_itertools-8.13.0-py3-none-any.whl (51 kB)
Requirement already satisfied: loguru<0.7,>=0.6 in c:\users\spencer\apps\python\lib\site-packages (from iscc-core) (0.6.0)
Collecting bitarray-hardbyte
  Using cached bitarray_hardbyte-2.3.8-cp39-cp39-win_amd64.whl (103 kB)
Collecting xxhash
  Using cached xxhash-3.0.0-cp39-cp39-win_amd64.whl (29 kB)
Collecting pybase64
  Using cached pybase64-1.2.2-cp39-cp39-win_amd64.whl (38 kB)
Collecting bases<0.3,>=0.2
  Using cached bases-0.2.1-py3-none-any.whl (35 kB)
Collecting data-url<2.0,>=1.0
  Using cached data_url-1.0.3-py3-none-any.whl
Collecting jcs<0.3,>=0.2
  Using cached jcs-0.2.1-py3-none-any.whl (7.6 kB)
Collecting blake3
  Using cached blake3-0.3.1-cp39-none-win_amd64.whl (193 kB)
Collecting uvarint<2.0,>=1.2
  Using cached uvarint-1.2.0-py3-none-any.whl (4.4 kB)
Requirement already satisfied: pydantic[dotenv] in c:\users\spencer\apps\python\lib\site-packages (from iscc-core) (1.9.1)
Collecting typing-validation
  Using cached typing_validation-0.0.1.post7-py3-none-any.whl (9.5 kB)
Requirement already satisfied: typing-extensions in c:\users\spencer\apps\python\lib\site-packages (from bases<0.3,>=0.2->iscc-core) (4.1.1)
Requirement already satisfied: win32-setctime>=1.0.0 in c:\users\spencer\apps\python\lib\site-packages (from loguru<0.7,>=0.6->iscc-core) (1.1.0)
Requirement already satisfied: colorama>=0.3.4 in c:\users\spencer\apps\python\lib\site-packages (from loguru<0.7,>=0.6->iscc-core) (0.4.4)
Collecting python-dotenv>=0.10.4
  Using cached python_dotenv-0.20.0-py3-none-any.whl (17 kB)
Building wheels for collected packages: iscc-core
  Building wheel for iscc-core (pyproject.toml) ... done
  Created wheel for iscc-core: filename=iscc_core-0.2.11-cp39-cp39-win_amd64.whl size=318612 sha256=d43418522914c53b2c2caf4dd37ce71f2a88ed7b370afe10b61285c943f9a06a
  Stored in directory: c:\users\spencer\appdata\local\pip\cache\wheels\98\b3\5c\fea4fbb7d772a0ba83d3402be2f7b6402ab12b60bba3d2edb0
Successfully built iscc-core
Installing collected packages: blake3, bitarray-hardbyte, xxhash, uvarint, typing-validation, python-dotenv, pybase64, more-itertools, jcs, data-url, base58, bases, iscc-core
Successfully installed base58-2.1.1 bases-0.2.1 bitarray-hardbyte-2.3.8 blake3-0.3.1 data-url-1.0.3 iscc-core-0.2.11 jcs-0.2.1 more-itertools-8.13.0 pybase64-1.2.2 python-dotenv-0.20.0 typing-validation-0.0.1.post7 uvarint-1.2.0 xxhash-3.0.0
```
